File: Tree/98validateBinarySearchTree.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def isValidBST(self, root):
        """
        :type root: TreeNode
        :rtype: bool
        """
        def dfs(root, less, top):
            if not root:
                return True
            if root.val >= less or root.val <= top:
                return False
            return dfs(root.left, min(root.val,less), top) and dfs(root.right, less, max(root.val,top))
        
        return dfs(root,float('inf'),float('-inf'))
            

File: Tree/98ValidateBinarySearchTree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    bool isValidBST(TreeNode* root) {
        return helper(root, LONG_MAX, LONG_MIN);
    }
    bool helper(TreeNode* root, long max, long min){
        if(!root) return true;
        if(root->val <= min || root->val >= max) return false;
        return helper(root->left, root->val, min) && helper(root->right, max, root->val);
    }
};